src/main/java/pl/agh/edu/master_diet/service/ProductBrowserService.java

```diff
@@ -0,0 +1,96 @@
+
+@Service
+@AllArgsConstructor(onConstructor = @__(@Autowired))
+public class ProductBrowserService {
+
+    private static final int ALLOW_SINGLE_TYPO_MINIMUM_WORD_LENGTH = 5;
+
+    private final ProductRepository productRepository;
+    private final RecentProductsRepository recentProductsRepository;
+    private final ConversionService conversionService;
+
+    public ProductSearchResponse searchProducts(String searchTerm, Integer pageIndex, Integer perPage) {
+        searchTerm = searchTerm.trim().toLowerCase();
+        List<Product> result = productRepository.findBySearchTerm(searchTerm);
+
+        if (searchTerm.length() >= ALLOW_SINGLE_TYPO_MINIMUM_WORD_LENGTH) {
+            List<Product> productsForSearchTermWithTypo = searchProductsForTermWithTypo(searchTerm);
+            productsForSearchTermWithTypo.removeAll(result);
+            result.addAll(productsForSearchTermWithTypo);
+        }
+
+        Integer maximumPageNumber = calculateMaximumPageNumber(result, perPage);
+        result = adjustListToPageInfo(result, pageIndex, perPage);
+
+        return ProductSearchResponse.builder()
+                .products(result.stream()
+                        .map(conversionService::convert)
+                        .collect(Collectors.toList()))
+                .maximumPageNumber(maximumPageNumber)
+                .build();
+    }
+
+    public GetRecentProductsResponse getRecentProducts(Integer pageIndex, Integer perPage, Long userId) {
+        List<RecentProduct> result = recentProductsRepository.findByUserId(userId);
+        Integer maximumPageNumber = calculateMaximumPageNumber(result, perPage);
+        result = adjustListToPageInfo(result, pageIndex, perPage);
+
+        return GetRecentProductsResponse.builder()
+                .products(result.stream()
+                        .map(conversionService::convert)
+                        .collect(Collectors.toList()))
+                .maximumPageNumber(maximumPageNumber)
+                .build();
+    }
+
+    private List<Product> searchProductsForTermWithTypo(String searchTerm) {
+        Set<Product> products = new LinkedHashSet<>();
+        for (int typoIndex = 0; typoIndex < searchTerm.length(); typoIndex++) {
+            String searchTermForTypo = prepareSearchTermForTypoAtIndex(searchTerm, typoIndex);
+            products.addAll(productRepository.findBySearchTerm(searchTermForTypo));
+        }
+
+        List<Product> result = new ArrayList<>(products);
+        result.sort((p1, p2) -> p2.getApprovals() - p1.getApprovals());
+        return result;
+    }
+
+    private String prepareSearchTermForTypoAtIndex(String searchTerm, int typoIndex) {
+        if (typoIndex == 0) {
+            return searchTerm.substring(1);
+        }
+        if (typoIndex == searchTerm.length() - 1) {
+            return searchTerm.substring(0, searchTerm.length() - 1);
+        }
+
+        return searchTerm.substring(0, typoIndex) + "_" + searchTerm.substring(typoIndex + 1);
```

Comment:
Changed method name and body.
`prepareSqlSearchPatternWithTypoAtSelectedIndex` takes search term and typo index and returns pattern for SQL _LIKE_ condition that takes this typo into account. For first and last character there is no need to use `_` wildcard as it will be surrounded by `%` anyway.
`searchProductsForTermWithTypo` is invoked when search term has minimal length that is defined in `ALLOW_SINGLE_TYPO_MINIMUM_WORD_LENGTH` constant. It creates patterns using `prepareSqlSearchPatternWithTypoAtSelectedIndex` method and returns list of products matching these patterns. For example it can return product with name `Sugar` even if `sugat` term was typed (letter case doesn't matter because of `LOWER` function included in JPQL query).